Add a list of download buttons for each episode to the LTI series list

Arnei opened the pull request:
Download button dropdown on the LTI series list

Adds a dropdown to each episode in a series. Clicking an entry in the dropdown 
will download the corresponding video file. @wsmirnow 
This pull request is intended as a quick change to r/8.x and will very likely be overwritten in r/9.x.

Expected result:
![DropDownOpen](https://user-images.githubusercontent.com/14070005/78668720-6ff94700-78db-11ea-91dc-e3a3f7e08684.jpg)


### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] be against the correct branch (features can only go into develop)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated testing
* [ ] have a clean commit history
* [ ] have proper commit messages (title and body) for all commits
* [ ] have appropriate tags applied

Commit:
Serieslist over LTI in Moodle

Commit:
Put the buttons in a dropdown

lkiesow commented:
Note that this likely conflicts with #1289

Commit:
Made download button wrap around in order to be visible on mobile resolutions

mtneug commented:
Not all institutions want to make it so easy to download the video files. Can this be made configurable?

Commit:
Added predicted file size to the download list

gregorydlogan commented:
This is also pointed at `develop`, which would be `9.x`.  You likely want to point this at `r/8.x`.

gregorydlogan commented on modules/runtime-info-ui/src/main/resources/ui/index.html:
```diff
@@ -79,7 +79,7 @@ <h2>Community Resources</h2>
             <li><a href="https://groups.google.com/a/opencast.org/forum/#!myforums">Mailing Lists (Google
                 Groups)</a></li>
-            <li><a href="http://webchat.freenode.net/?channels=opencast">IRC-Channel (Webinterface)</a></li>
+            <li><a href="http://webchat.freenode.net/?channels=opencast">IRC-Channel (Webinterfafa face)</a></li>
```
This should be reverted

lkiesow commented:
> This is also pointed at `develop`, which would be `9.x`. You likely want to point this at `r/8.x`.

Not sure if we really want to bring this in before #1289? This should at least go by the release managers (@karendolan, @rrolf). We are not only putting a feature in 8.x this time but deliberately breaking another pull request (we do that, correct? I'm actually just assuming that.). But then… it's not my pull request so I care less then I would otherwise ;-P

gregorydlogan commented:
IMO #1289 takes priority since its larger and been around too long. I have it built, was going to look at it in the next couple of days.

I was only commenting on the branch target since the description mentions 8.x. I don't really care where this goes, or if it goes in at all :)

LukasKalbertodt commented on etc/ui-config/mh_default_org/ltitools/series.css:
```diff
@@ -50,6 +58,58 @@ a.episode div.desc h2 {
+  margin-top: 10px;
+  margin-right: 10px;
+  margin-left: 10px;
+  margin-bottom: 10px;
```
Why not just `margin: 10px`?

LukasKalbertodt commented on etc/ui-config/mh_default_org/ltitools/series.css:
```diff
@@ -50,6 +58,58 @@ a.episode div.desc h2 {
 }
 
+/* Style The Dropdown Button */
+.dropbtn {
```
I personally don't think that shortening words like "button" to "btn" is useful at all. Why not simply `drop-button`?

LukasKalbertodt commented on etc/ui-config/mh_default_org/ltitools/series.css:
```diff
@@ -50,6 +58,58 @@ a.episode div.desc h2 {
+}
+
+/* Change color of dropdown links on hover */
+.dropdown-content a:hover {background-color: #f1f1f1}
```
This should be 3 lines. I don't think we gain anything by squashing this into one.

LukasKalbertodt commented on modules/lti/src/main/resources/tools/series/index.html:
```diff
@@ -23,14 +23,27 @@
+      <div class=preview><img alt=Preview src={{image}}></div>
+      <div class=desc>
+        <h2>{{title}}</h2>
+        {{i18ncreator}}<br>
```
Here the `<br />` changed to `<br>`. I don't think it's problematic, I'm just not sure if it's intended.

LukasKalbertodt commented on modules/lti/src/main/resources/tools/series/index.html:
```diff
@@ -23,14 +23,27 @@
+      <button class=dropbtn>Direct Download</button>
+      <div class=dropdown-content>
+      {{#download}}
+        <a href={{path-to-file}} download="">
```
Isn't there some useful filename we could add to `download=`? If not, I think you can simply remove the `=""` part (`<a ... download>`)

Arnei replied:
I suppose you could name the file similarly to the download button name, but I do not know whether this is desirable or not.

LukasKalbertodt replied:
It looks like the URL contains a somewhat useful filename. So I guess we can just use `download` without `=""`.

LukasKalbertodt commented on modules/lti/src/main/resources/tools/series/index.html:
```diff
@@ -23,14 +23,27 @@
+      <div class=dropdown-content>
+      {{#download}}
+        <a href={{path-to-file}} download="">
+        {{{download-button-name}}}
```
I think this line should be indented one additional level?

LukasKalbertodt commented on modules/lti/src/main/resources/tools/series/index.html:
```diff
@@ -23,14 +23,27 @@
+        <a href={{path-to-file}} download="">
+        {{{download-button-name}}}
+        </a>
+        <br>
```
Is this line break really necessary? The `<a>` elements are already `display: block`, right? If it's about spacing between the links, I'd probably rather solve this via `margin-bottom` on the `<a>` elements. 

Additionally, it seems that this is creating an empty space after the last element which is undesirable. You can fix this in CSS like this:

```css
.dropdown-content a:not(:last-child) {
    margin-bottom: ...;
}
```

LukasKalbertodt commented on modules/lti/src/main/resources/tools/series/series.js:
```diff
@@ -68,6 +68,12 @@ function getSeries() {
 
+function capitalize(s) {
+  if (typeof s !== 'string')
+    return '';
```
I'm not sure about the current policy but I remember the community discussing this on the mailing list. I *think* the outcome was that we want to enforce adding braces `{ }` in these places. I personally absolutely think that braces here should be required.

LukasKalbertodt commented on modules/lti/src/main/resources/tools/series/series.js:
```diff
@@ -115,6 +121,32 @@ function loadPage(page) {
+      tracks = Array.isArray(tracks) ? tracks : [tracks];
+      for (var k = 0; k < tracks.length; k++)
+      {
+        if( tracks[k].url.endsWith('mp4') || tracks[0].url.endsWith('webm')) {
```
Spacing is inconsistent. From looking at the rest of this file, this should probably be

```js
        if (tracks[k].url.endsWith('mp4') || tracks[0].url.endsWith('webm')) {
```

Also: why only mp4 and webm? 

Also also: `tracks[0]`... is that intended or should it be `tracks[k]`? If the latter, the `for` loop should be

```js
for (const track of tracks) {
```

Arnei replied:
Checking whether the url of a track points to a mp4 or webm file was recommended as a way to check whether the track is for a downloadable video or not.

And indeed, it should be `tracks[k]`.

LukasKalbertodt replied:
If webm and mp4 is the only thing produces by Opencast, then that's fine. I am just not sure about that.

LukasKalbertodt commented on modules/lti/src/main/resources/tools/series/series.js:
```diff
@@ -115,6 +121,32 @@ function loadPage(page) {
+        if( tracks[k].url.endsWith('mp4') || tracks[0].url.endsWith('webm')) {
+
+          // Calculate filesize
+          var fileSize = (tracks[k].video.bitrate / 8 / 1000 / 1000)      // Converts to byte, kilobyte, megabyte /s
+            * (tracks[k].video.framecount / tracks[k].video.framerate);
```
Is there no way to get access to the actual file size? This seems like a ... hacky approximation?

lkiesow replied:
This also disregards the audio.
If the size is not in the JSON, it probably makes sense to not display it.
And I know that I originally suggested to show the size :D

Arnei replied:
It may seem that way because it is ^^'. The JSON containing the track information does not specify a file size, so I thought I'd use the other information at hand. Not sure if I can get the file size from the url itself, something to look into.

Arnei replied:
So apparently you can just _use a HTTP HEAD request and then look at Content-Length response header. ( [https://bitexperts.com/Question/Detail/3316/determine-file-size-in-javascript-without-downloading-a-file](url) )

Unsure whether it is worth it to send a bunch of web requests, since the hacky approximation seems to arrive at fairly similar numbers.

LukasKalbertodt replied:
> If the size is not in the JSON, it probably makes sense to not display it.

I guess I agree with that. Doesn't seem like a critical feature, so just remove the file size?

Arnei replied:
Huh, I figured the bitrate would include the audio. 
But if showing file size is not that important anyway, I'll cut it.

LukasKalbertodt commented on modules/lti/src/main/resources/tools/series/series.js:
```diff
@@ -115,6 +121,32 @@ function loadPage(page) {
+
+          // Collect output to mustache
+          buttonData.push ({
+            'download-button-name' : capitalize( tracks[k].type.split('/')[0] )
+            + '<br /> ('
+            + tracks[k].video.resolution
+            + ', '
+            + fileSize.toFixed(2) + 'MB'
+            + ')',
```
I don't think this is how it should be done. Rather than adding HTML in this string and then using `{{{ x }}}` to not escape the string, you should use two strings (`download-button-name` and `download-description` or something like that). Using `{{{ }}}` in Mustache should be avoided if possible.

LukasKalbertodt reviewed (changes requested):
I added a couple of line comments. I tested locally and it seems to work for me. If you have any questions about my comments, just ask!

Commit:
Cleaned up the html and css according to Lukas' recommendations

Commit:
Cleaned up the js and html according to Lukas' recommendations.
Removed approximative file size display.
Fixed incorrect for-loop index by switching to foreach.

Commit:
Fixed typo

Arnei commented:
The pull request is too specific to be merged into any main branch.
